cl/alerts/tasks.py

```diff
@@ -28,117 +30,142 @@ def enqueue_docket_alert(d_pk: int) -> bool:
     return create_redis_semaphore("ALERTS", key, ttl=60 * 10)
 
 
+def get_docket_alert_recipients(
+    d_pk: int,
+    recap_email_recipients: list[str] = [],
+) -> tuple[
+    list[tuple[str, str]],
+    list[tuple[str, str]],
+    list[tuple[str, str]],
+    list[str],
+]:
+    """Get the notification's recipients for a docket alert.
+    :param d_pk: Docket primary key
+    :param recap_email_recipients: List of @recap.email addresses to send the
+    notification to.
+    :return: A Tuple of lists of tuples (email_address, secret_key) and a list
+    of @recap.email addresses that don't belong to any user if any.
+    """
+
+    # Lists of tuples (email_address, docket_alert.secret_key)
+    docket_alert_recipients_list = []
+    recap_email_first_time_auto_subscribe_list = []
+    recap_email_first_time_no_auto_subscribe_list = []
+    # List of @recap.email addresses that don't belong to any user
+    recap_email_user_does_not_exist_list = []
+
+    # First, get current docket alert recipients to avoid duplicate alerts
+    docket_alerts_current_subscribers = DocketAlert.objects.filter(
+        docket_id=d_pk, alert_type=DocketAlert.SUBSCRIPTION
+    )
+    for da in docket_alerts_current_subscribers:
+        recipient_tuple = (da.user.email, da.secret_key)
+        docket_alert_recipients_list.append(recipient_tuple)
+
+    # Get recap email recipients and create new docket alerts objects
+    for email_address in recap_email_recipients:
+        try:
+            user_profile = UserProfile.objects.get(recap_email=email_address)
+        except UserProfile.DoesNotExist:
+            recap_email_user_does_not_exist_list.append(email_address)
+            continue
+        email_address = user_profile.user.email
+        docket_alert_exist = DocketAlert.objects.filter(
+            docket_id=d_pk, user=user_profile.user
+        ).exists()
+        if docket_alert_exist:
+            # If a docket alert exists for this @recap.email user, avoid
+            # sending the first email
+            continue
+        if user_profile.auto_subscribe:
+            # First time recipient, auto_subscribe True
+            docket_alert = DocketAlert.objects.create(
+                docket_id=d_pk, user=user_profile.user
+            )
+            recipient_tuple = (email_address, docket_alert.secret_key)
+            recap_email_first_time_auto_subscribe_list.append(recipient_tuple)
+        else:
+            # First time recipient, auto_subscribe False
+            docket_alert = DocketAlert.objects.create(
+                docket_id=d_pk,
+                user=user_profile.user,
+                alert_type=DocketAlert.UNSUBSCRIPTION,
+            )
+            recipient_tuple = (email_address, docket_alert.secret_key)
+            recap_email_first_time_no_auto_subscribe_list.append(
```

Comment:
This `objects.create(` code is almost completely duplicated. I think when you make the refactor to return a list of dicts/objects, this should simplify. It should wind up with something like:

```
    # Somewhere above: md = MessageData()
    alert_type = DocketAlert.SUBSCRIPTION if auto_subscribe else DocketAlert.UNSUBSCRIPTION
    docket_alert = DocketAlert.objects.create(
        docket_id=d_pk,
        user=user_profile.user,
        alert_type=alert_type,
    )
    md.email_address = email_address
    md.secret_key = docket_alert.secret_key
    md.auto_subscribe = user_profile.auto_subscribe
    message.append(md)
return messages
```

Reply:
Yeah, this simplification was possible thanks to the new data class.